File: packages/device_protocol/src/device_protocol/topics.py
from __future__ import annotations

COMMAND_TOPIC_FMT = "pots/{pot_id}/command"
SENSORS_TOPIC_FMT = "pots/{pot_id}/sensors"
STATUS_TOPIC_FMT = "pots/{pot_id}/status"

LEGACY_FIRMWARE_TELEMETRY_FILTER = "projectplant/pots/+/telemetry"
LEGACY_FIRMWARE_STATUS_FILTER = "projectplant/pots/+/status"
CANONICAL_COMMAND_FILTER = "pots/+/command"
CANONICAL_SENSOR_TOPIC_FMT = SENSORS_TOPIC_FMT
CANONICAL_SENSOR_FILTER = "pots/+/sensors"
CANONICAL_STATUS_TOPIC_FMT = STATUS_TOPIC_FMT
CANONICAL_STATUS_FILTER = "pots/+/status"
DEVICE_STATE_FILTER = "plant/+/state"
PLANT_TELEMETRY_FILTER = "plant/+/telemetry"
METRICS_TOPIC_FMT = "plant/{plant_id}/et/metrics"
IRRIGATION_CMD_TOPIC_FMT = "plant/{plant_id}/irrigation/cmd"
MQTT_PING_TOPIC = "lab/ping"


def command_topic(pot_id: str) -> str:
    return COMMAND_TOPIC_FMT.format(pot_id=pot_id)


def sensor_topic(pot_id: str) -> str:
    return SENSORS_TOPIC_FMT.format(pot_id=pot_id)


def status_topic(pot_id: str) -> str:
    return STATUS_TOPIC_FMT.format(pot_id=pot_id)


def legacy_telemetry_topic(pot_id: str) -> str:
    return f"projectplant/pots/{pot_id}/telemetry"


def legacy_status_topic(pot_id: str) -> str:
    return f"projectplant/pots/{pot_id}/status"


def device_state_topic(device_id: str) -> str:
    return f"plant/{device_id}/state"


def plant_telemetry_topic(plant_id: str) -> str:
    return f"plant/{plant_id}/telemetry"


def et_metrics_topic(plant_id: str) -> str:
    return METRICS_TOPIC_FMT.format(plant_id=plant_id)


def irrigation_command_topic(plant_id: str) -> str:
    return IRRIGATION_CMD_TOPIC_FMT.format(plant_id=plant_id)

File: packages/protocol/src/topics.js
export const CANONICAL_TOPIC_PREFIX = "pots";
export const LEGACY_FIRMWARE_PREFIX = "projectplant/pots";
export const DEVICE_STATE_FILTER = "plant/+/state";
export const PLANT_TELEMETRY_FILTER = "plant/+/telemetry";
export const MQTT_PING_TOPIC = "lab/ping";
export const CANONICAL_SENSOR_FILTER = "pots/+/sensors";
export const CANONICAL_STATUS_FILTER = "pots/+/status";
export const CANONICAL_COMMAND_FILTER = "pots/+/command";
export const LEGACY_FIRMWARE_TELEMETRY_FILTER = "projectplant/pots/+/telemetry";
export const LEGACY_FIRMWARE_STATUS_FILTER = "projectplant/pots/+/status";
export function potTopic(potId, kind) {
    assertTopicId(potId, "potId");
    return `${CANONICAL_TOPIC_PREFIX}/${potId}/${kind}`;
}
export function sensorTopic(potId) {
    return potTopic(potId, "sensors");
}
export function statusTopic(potId) {
    return potTopic(potId, "status");
}
export function commandTopic(potId) {
    return potTopic(potId, "command");
}
export function legacyTelemetryTopic(potId) {
    assertTopicId(potId, "potId");
    return `${LEGACY_FIRMWARE_PREFIX}/${potId}/telemetry`;
}
export function legacyStatusTopic(potId) {
    assertTopicId(potId, "potId");
    return `${LEGACY_FIRMWARE_PREFIX}/${potId}/status`;
}
export function deviceStateTopic(deviceId) {
    assertTopicId(deviceId, "deviceId");
    return `plant/${deviceId}/state`;
}
export function plantTelemetryTopic(plantId) {
    assertTopicId(plantId, "plantId");
    return `plant/${plantId}/telemetry`;
}
export function etMetricsTopic(plantId) {
    assertTopicId(plantId, "plantId");
    return `plant/${plantId}/et/metrics`;
}
export function irrigationCommandTopic(plantId) {
    assertTopicId(plantId, "plantId");
    return `plant/${plantId}/irrigation/cmd`;
}
export function parseCanonicalTopic(topic) {
    const parts = topic.split("/");
    if (parts.length !== 3) {
        return null;
    }
    const [prefix, potId, kind] = parts;
    if (prefix !== CANONICAL_TOPIC_PREFIX || !potId) {
        return null;
    }
    if (kind === "sensors" || kind === "status" || kind === "command") {
        return { namespace: "canonical", potId, kind };
    }
    return null;
}
export function parseSensorTopic(topic) {
    const parsed = parseCanonicalTopic(topic);
    return parsed?.kind === "sensors" ? parsed.potId : null;
}
export function parseLegacyFirmwareTopic(topic) {
    const parts = topic.split("/");
    if (parts.length !== 4) {
        return null;
    }
    const [project, pots, potId, kind] = parts;
    if (project !== "projectplant" || pots !== "pots" || !potId) {
        return null;
    }
    if (kind === "telemetry" || kind === "status") {
        return { namespace: "legacy-firmware", potId, kind };
    }
    return null;
}
export function parseDeviceStateTopic(topic) {
    const parts = topic.split("/");
    if (parts.length === 3 && parts[0] === "plant" && parts[2] === "state" && parts[1]) {
        return { namespace: "device-state", deviceId: parts[1] };
    }
    return null;
}
export function parsePlantTopic(topic) {
    const parts = topic.split("/");
    if (parts.length < 3 || parts[0] !== "plant" || !parts[1]) {
        return null;
    }
    if (parts.length === 3 && parts[2] === "telemetry") {
        return { namespace: "plant", plantId: parts[1], kind: "telemetry" };
    }
    if (parts.length === 4 && parts[2] === "et" && parts[3] === "metrics") {
        return { namespace: "plant", plantId: parts[1], kind: "et/metrics" };
    }
    if (parts.length === 4 && parts[2] === "irrigation" && parts[3] === "cmd") {
        return { namespace: "plant", plantId: parts[1], kind: "irrigation/cmd" };
    }
    return null;
}
export function parseTopic(topic) {
    return (parseCanonicalTopic(topic) ??
        parseLegacyFirmwareTopic(topic) ??
        parseDeviceStateTopic(topic) ??
        parsePlantTopic(topic));
}
function assertTopicId(value, label) {
    if (!value || value.includes("/") || value.includes("+") || value.includes("#")) {
        throw new Error(`${label} is required and must be a single MQTT topic segment`);
    }
}
